### PR TITLE
client ID

### PR DESCRIPTION
**Pull request by deltics**
## Summary by CodeRabbit

- **New Features**
  - Introduced HTTP GET request functionality in the `Returning` struct, enhancing JSON handling capabilities.

- **Refactor**
  - Renamed client identification fields from `name` to `id` across various client and mockClient files for consistency.

- **Tests**
  - Updated test cases to reflect the new client identification name changes from `name` to `id`.

- **Chores**
  - Added `.coderabbit.yml` configuration file for YAML language server settings.

### CODE REVIEW

**Comment by deltics** [comment]
@bito /review

**Comment by deltics** [comment]
<div>
<h2>Code Review Agent Run #3867f3</h2>
<div>
<ul>

<li>
<b>AI Based Review: </b><small>✔️</small>&nbsp;Successful
</li>

</ul>
</div>
<div>
<h3>Code Review Overview</h3>
<ul>
<li><b>Summary</b>: The changes primarily focus on replacing 'name' with 'id' for client identification, adding sequence increments for unique client IDs, and updating related methods and tests. A new configuration file '.coderabbit.yml' has been added with various settings. Additionally, a new file 'requester.go' introduces a Returning struct and a Get method for HTTP requests.</li>
<li><b>Files</b>: 8</li>
<li><b>Issue found</b>: Total - 10, High importance - 10 &nbsp;&nbsp;&nbsp;&nbsp; <a href="https://github.com/blugnu/http/pull/3/files">See detailed feedback -></a></li>
<li><b>Code change type</b>: Functionality, Documentation, Error Handling, Scalability</li>
<li><b>Unit tests added</b>: True</li>
<li><b>Estimated effort to review (1-5, lower is better)</b>: 3</li>
</ul>
</div>
<div>
<h3>High-level Feedback</h3>
Ensure that all client IDs are unique and thread-safe to avoid race conditions. Improve documentation clarity by providing detailed comments and examples. Handle HTTP error status codes appropriately in the 'Get' method. Add comprehensive test cases to cover edge scenarios and potential errors. Validate and sanitize inputs to prevent security risks and ensure robust error handling.
</div>
<br>

<div>
<h3>Detailed Feedback</h3>

<div id="issue">
<div>
:page_facing_up: mockClient_test.go
</div>
<details>
<summary>
Issues: Total - 1, High importance - 1
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211822">Line 39-39</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

<div id="issue">
<div>
:page_facing_up: requester.go
</div>
<details>
<summary>
Issues: Total - 1, High importance - 1
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211826">Line 14-14</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

<div id="issue">
<div>
:page_facing_up: clientOptions.go
</div>
<details>
<summary>
Issues: Total - 3, High importance - 3
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211828">Line 13-13</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211835">Line 30-30</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211849">Line 32-34</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

<div id="issue">
<div>
:page_facing_up: client.go
</div>
<details>
<summary>
Issues: Total - 1, High importance - 1
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211850">Line 91-91</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

<div id="issue">
<div>
:page_facing_up: clientOptions_test.go
</div>
<details>
<summary>
Issues: Total - 2, High importance - 2
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211851">Line 28-40</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211852">Line 117-129</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

<div id="issue">
<div>
:page_facing_up: client_test.go
</div>
<details>
<summary>
Issues: Total - 1, High importance - 1
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211853">Line 144-144</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

<div id="issue">
<div>
:page_facing_up: mockClient.go
</div>
<details>
<summary>
Issues: Total - 1, High importance - 1
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/blugnu/http/pull/3/files#r1667211854">Line 211-211</a>
</td>
<td width="70%">
 <small>🔴</small> High importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

</div>




<span><i>AI Code Review powered by</i> <sub><sub><a href="https://bito.ai/" target="_blank">Bito Logo</a></sub></sub></span>

</div>

File: mockClient_test.go [hunk 1]
```diff
@@ -36,11 +36,11 @@ func TestNewMockClient(t *testing.T) {
 
 	// ASSERT
 	if c, ok := test.IsType[client](t, c); ok {
-		test.That(t, c.name).Equals("foo")
+		test.That(t, c.id).Equals("foo")
```
Comment (deltics) [inline mockClient_test.go:39]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test assertion 'test.That(t, c.id).Equals("foo")' does not verify if the 'id' field is correctly set in the 'client' struct. If the 'id' field is not set correctly, the test will fail, but it won't provide any information about the actual value of 'id'.
</div>
<div id="fix">
<b>Fix:</b> Add a message to the assertion to provide more context if the test fails. This will help in debugging by showing the actual value of 'id'.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -39,7 +39,7 @@
  if c, ok := test.IsType[client](t, c); ok {
 -    test.That(t, c.id).Equals("foo")
 +    test.That(t, c.id).Equals("foo", "Expected client id to be 'foo', but got: %v", c.id)
      test.That(t, c.url).Equals("mock://hostname")
  }
```
</div>
</details>
</div>
<br>


</div>

File: requester.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/blugnu/errorcontext"
+	"github.com/blugnu/http/request"
+)
+
+type Returning[T any] struct {
+	HttpClient
+}
+
+func (r Returning[T]) Get(ctx context.Context, path string, opts ...func(*Request) error) (*T, error) {
```
Comment (deltics) [inline requester.go:14]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The 'Get' method in the 'Returning' struct does not handle HTTP status codes that indicate an error (e.g., 4xx or 5xx status codes). This can lead to incorrect assumptions about the success of the request.
</div>
<div id="fix">
<b>Fix:</b> Check the HTTP status code of the response and handle error status codes appropriately.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -16,6 +16,11 @@ func (r Returning[T]) Get(ctx context.Context, path string, opts ...func(*Request) error) (*T, error) {
  	resp, err := r.HttpClient.Get(ctx, path, opts...)
  	if err != nil {
  		return nil, errorcontext.Wrap(ctx, err)
 +	}
 +	if resp.StatusCode >= 400 {
 +		return nil, errorcontext.Errorf(ctx, "HTTP request failed with status code %d", resp.StatusCode)
 +	}
 +	defer resp.Body.Close()
 
  	result, err := UnmarshalJSON[T](resp)
  	if err != nil {
```
</div>
</details>
</div>
<br>


</div>

File: clientOptions.go [hunk 3]
```diff
@@ -6,6 +6,17 @@ import (
 	"net/url"
 )
 
+// ClientID sets the client ID for requests made using the client.  The client ID is
+// typically used to differentate between clients in log entries.  If no client ID is
+// provided a default value will be applied consisting of the string "http-<seq>"
+// where <seq> is a number that increments for each client created.
+func ClientID(id string) ClientOption {
```
Comment (deltics) [inline clientOptions.go:13]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The 'ClientID' function does not handle the case where the provided 'id' is an empty string. This could lead to clients being created with an empty ID, which might cause issues in logging or client differentiation.
</div>
<div id="fix">
<b>Fix:</b> Add a check to ensure that the 'id' is not an empty string. If it is, assign a default value.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -13,7 +13,11 @@ func ClientID(id string) ClientOption {
 	return func(c *client) error {
 		if id == "" {
 			id = fmt.Sprintf("http-%d", nextClientID())
 		}
 		c.id = id
 		return nil
 	}
 }
```
</div>
</details>
</div>
<br>


</div>

File: clientOptions.go [hunk 4]
```diff
@@ -16,11 +27,11 @@ func MaxRetries(n uint) ClientOption {
 }
 
 // URL sets the base URL for requests made using the client.  The URL may be specified
-// as a string or a *url.URL.
-//
-// If a string is provided, it will be parsed to ensure it is a valid, absolute URL.
+// as any of:
```
Comment (deltics) [inline clientOptions.go:30]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The removal of detailed comments explaining the expected input types and parsing behavior for a URL reduces the clarity and documentation quality of the code. This can lead to confusion for future developers who may not understand the expected input types and how they are handled.
</div>
<div id="fix">
<b>Fix:</b> Restore the detailed comments explaining the expected input types and parsing behavior for a URL. This will ensure that future developers have a clear understanding of how the input should be provided and how it will be processed.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -19,3 +30,5 @@
 -// as a string or a *url.URL.
 -//
 -// If a string is provided, it will be parsed to ensure it is a valid, absolute URL.
 +// as any of:
 +// - a string, which will be parsed to ensure it is a valid, absolute URL.
 +// - a *url.URL, which will be used directly.
```
</div>
</details>
</div>
<br>


</div>

File: clientOptions.go [hunk 5]
```diff
@@ -16,11 +27,11 @@ func MaxRetries(n uint) ClientOption {
...
 //
-// If a URL is provided is must be absolute.
+//	string        // a which parses to a valid, absolute URL
+//	url.URL       // a valid, absolute URL
+//	*url.URL      // a valid, absolute URL
```
Comment (deltics) [inline clientOptions.go:34]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The new comments added are not clear and do not provide sufficient context or explanation. They seem to be incomplete and may confuse future developers.
</div>
<div id="fix">
<b>Fix:</b> Provide a more detailed and clear explanation of the different URL types that can be used, including examples if possible.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -32,3 +32,6 @@
 +// The following types are supported for URLs:
 +// - string: a string that parses to a valid, absolute URL
 +// - url.URL: a valid, absolute URL
 +// - *url.URL: a pointer to a valid, absolute URL
```
</div>
</details>
</div>
<br>


</div>

File: client.go [hunk 6]
```diff
@@ -80,9 +85,10 @@ type client struct {
 // The url typically includes the protocol, hostname and port for the client
 // but may include any additional url components consistently required for
 // requests performed using the client.
-func NewClient(name string, opts ...ClientOption) (HttpClient, error) {
+func NewClient(opts ...ClientOption) (HttpClient, error) {
+	seq++
 	w := client{
-		name:    name,
+		id:      "http-" + strconv.Itoa(seq),
```
Comment (deltics) [inline client.go:91]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The current ID generation logic using a global sequence counter (seq) may lead to race conditions in a concurrent environment, causing duplicate IDs or other inconsistencies.
</div>
<div id="fix">
<b>Fix:</b> Use a thread-safe mechanism to generate unique client IDs, such as using a UUID generator or a sync/atomic counter.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -88,7 +88,7 @@
  func NewClient(opts ...ClientOption) (HttpClient, error) {
      seq++
      w := client{
 -        id:      "http-" + strconv.Itoa(seq),
 +        id:      "http-" + uuid.New().String(),
          wrapped: http.DefaultClient,
      }
      errs := make([]error, 0, len(opts))
      for _, opt := range opts {
          if err := opt(&w); err != nil {
              errs = append(errs, err)
          }
      }
      if len(errs) > 0 {
          return nil, fmt.Errorf("%w: %w", ErrInitialisingClient, errors.Join(errs...))
      }
      return w, nil
  }
```
</div>
</details>
</div>
<br>


</div>

File: clientOptions_test.go [hunk 7]
```diff
@@ -25,6 +25,19 @@ func TestClientOptions(t *testing.T) {
...
+				sut := &client{}
+
+				// ACT
+				err := ClientID("foo")(sut)
+
+				// ASSERT
+				test.That(t, err).IsNil()
+				test.That(t, sut).Equals(&client{id: "foo"})
+			},
+		},
```
Comment (deltics) [inline clientOptions_test.go:40]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case for 'ClientID' does not cover edge cases such as passing an empty string or a very long string as the client ID. This could lead to potential issues if the 'ClientID' function does not handle these cases properly.
</div>
<div id="fix">
<b>Fix:</b> Add additional test cases to cover edge cases for the 'ClientID' function, such as passing an empty string and a very long string.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -40,0 +41,20 @@
 +		{scenario: "ClientID/empty",
 +			exec: func(t *testing.T) {
 +				// ARRANGE
 +				sut := &client{}
 +
 +				// ACT
 +				err := ClientID("")(sut)
 +
 +				// ASSERT
 +				test.That(t, err).IsNil()
 +				test.That(t, sut).Equals(&client{id: ""})
 +			},
 +		},
 +		{scenario: "ClientID/long",
 +			exec: func(t *testing.T) {
 +				// ARRANGE
 +				sut := &client{}
 +				longID := strings.Repeat("a", 1000)
 +
 +				// ACT
 +				err := ClientID(longID)(sut)
 +
 +				// ASSERT
 +				test.That(t, err).IsNil()
 +				test.That(t, sut).Equals(&client{id: longID})
 +			},
 +		},
```
</div>
</details>
</div>
<br>


</div>

File: clientOptions_test.go [hunk 8]
```diff
@@ -96,6 +109,33 @@ func TestClientOptions(t *testing.T) {
...
+				client := &client{}
+				url, _ := url.Parse("example.com")
+
+				// ACT
+				err := URL(*url)(client)
+
+				// ASSERT
+				test.Error(t, err).Is(ErrInvalidURL)
+			},
+		},
```
Comment (deltics) [inline clientOptions_test.go:129]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case for 'URL/*URL/relative' does not handle the potential panic that could occur if 'url.Parse' returns an error. This could lead to a test failure due to an unhandled error.
</div>
<div id="fix">
<b>Fix:</b> Add error handling for the 'url.Parse' function to ensure that the test does not panic if an error occurs.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -121,0 +122,2 @@
 +				if err != nil {
 +					t.Fatalf("failed to parse URL: %v", err)
 +				}
```
</div>
</details>
</div>
<br>


</div>

File: client_test.go [hunk 9]
```diff
@@ -140,6 +140,10 @@ func TestNewRequest(t *testing.T) {
 	}
 	for _, tc := range testcases {
 		t.Run(tc.scenario, func(t *testing.T) {
+			// ARRANGE
+			seq = 0
```
Comment (deltics) [inline client_test.go:144]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The variable 'seq' is being set to 0 without any context or explanation. This could lead to confusion or errors if 'seq' is used elsewhere in the test cases or if its value is expected to be different.
</div>
<div id="fix">
<b>Fix:</b> Provide a clear explanation or comment on why 'seq' is being reset to 0. Ensure that this reset does not interfere with other test cases or the overall test logic.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -143,4 +143,6 @@
 			// ARRANGE
 +			// Resetting seq to 0 to ensure a consistent starting point for the sequence in each test case
 			seq = 0
			
 			// ACT
```
</div>
</details>
</div>
<br>


</div>

File: mockClient.go [hunk 10]
```diff
@@ -207,7 +208,7 @@ func (mock mockClient) ExpectationsWereMet() error {
 func (mock *mockClient) Expect(method string, path string) *MockRequest {
 	if mock.next > 0 {
 		msg := "requests have already been made"
-		panic(fmt.Errorf("%s: %w: %s", mock.name, ErrCannotChangeExpectations, msg))
+		panic(fmt.Errorf("%s: %w: %s", mock.id, ErrCannotChangeExpectations, msg))
```
Comment (deltics) [inline mockClient.go:211]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/blugnu/http/pull/3#issuecomment-2211525124>#3867f3</a><span> - 07/06/2024, 12:12 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The panic message uses 'mock.id' which is a change from 'mock.name'. If 'mock.id' is not properly initialized or if it is an empty string, the error message might not be informative enough to debug the issue.
</div>
<div id="fix">
<b>Fix:</b> Ensure that 'mock.id' is properly initialized and contains a meaningful value before it is used in the panic message. If 'mock.id' can be empty or uninitialized, consider adding a check or a default value to make the error message more informative.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -209,6 +209,10 @@
  func (mock *mockClient) Expect(method string, path string) *MockRequest {
      if mock.next > 0 {
          msg := "requests have already been made"
 +        if mock.id == "" {
 +            mock.id = "unknown-client"
 +        }
          panic(fmt.Errorf("%s: %w: %s", mock.id, ErrCannotChangeExpectations, msg))
      }
```
</div>
</details>
</div>
<br>


</div>

File: mockClient_test.go (hunk 1 above)
Comment (deltics) [reply mockClient_test.go:39]: The blugnu/test package used for this test uses a fluent API to assert expected vs actual values.  If the test fails, the value of `c.id` and the value in the argument passed to the `Equals()` fluent method are both included in the test failure report.
